zebra-consensus/src/checkpoint.rs

```diff
@@ -754,16 +795,74 @@ impl Service<Arc<Block>> for CheckpointVerifier {
         // on the next call(). Failures always reject a block, so we know
         // there will be at least one more call().
         //
-        // TODO(teor): retry on failure (low priority, failures should be rare)
-        self.process_checkpoint_range();
+        // We don't retry with a smaller range on failure, because failures
+        // should be rare.
+        let valid_blocks = self.process_checkpoint_range();
 
         metrics::gauge!("checkpoint.queued_slots", self.queued.len() as i64);
 
-        async move {
-            // Remove the Result<..., RecvError> wrapper from the channel future
-            rx.await
-                .expect("CheckpointVerifier does not leave dangling receivers")
+        // Optimisation: return a small future in the common case.
+        //
+        // Rust 1.46 and later have known regressions in future size and
+        // complexity, so we want to return a small future in the common case.
```

Comment:
My preference would be to not change our code to optimize for this regression -- there's much more engineering effort applied to Rust than to our project, so my expectation is that the regression will be fixed and we'll then have to update our code, or more likely, we'll forget or be busy and end up with an unneeded optimization.

I think optimizing for this kind of thing only makes sense when (1) the code is functionally complete and (2) we have some measurement that it's a real performance impact.

Reply:
Would be avoided by the suggestion in https://github.com/ZcashFoundation/zebra/pull/1012#discussion_r485181158